### ISSUE TITLE
EVMS: Integrate core framework

### ISSUE THREAD

**Issue by Kishanrajput** [JeffersonLab/jlab_datascience_core#5]

**Comment by dlersch** [comment]
Started implementing an image to numpy parser. All this parser does, is loading .png files and combining them into one .npy file. Attached is a screenshot of the unit-test. The unit-test just does an dimensional check, i.e. tests if the output dimensions of the numpy array math the user settings.
<img width="741" alt="utest_image2numpy_parser" src="https://github.com/JeffersonLab/jlab_datascience_core/assets/19531538/33c287fd-f898-487c-be37-edc69d665877">

**Comment by dlersch** [comment]
Added simple linear scaler which transforms a given numpy array X according to: A*X + B where the constants A,B are defined in the user config. The unit-test creates a numpy array, runs the transformation and reverses it. Dimensional checks ensure that the unit-test was successful. Attached is a screenshot of the unit-test
<img width="686" alt="utest_numpy_linear_scaler" src="https://github.com/JeffersonLab/jlab_datascience_core/assets/19531538/a6d35b81-0e44-4335-8664-6a0de4ad9ea0">

**Comment by dlersch** [comment]
Started working on the module module by implementing a [keras CNN Autoencoder](https://github.com/JeffersonLab/jlab_datascience_core/blob/5-evms-integrate-core-framework/jlab_datascience_toolkit/model/keras_cnn_ae.py). The model can be run with / without a [keras early stopping callback](https://github.com/JeffersonLab/jlab_datascience_core/blob/5-evms-integrate-core-framework/jlab_datascience_toolkit/utils/keras_callbacks/keras_early_stopping.py). The corresponding [unit-test](https://github.com/JeffersonLab/jlab_datascience_core/blob/5-evms-integrate-core-framework/utests/utest_keras_cnn_ae.py) trains the AE on a fraction of the mnist data which can be called via this [helper class](https://github.com/JeffersonLab/jlab_datascience_core/blob/5-evms-integrate-core-framework/jlab_datascience_toolkit/utils/get_mnist.py).  Shown below are the plots that are produced by the unit-test. Please note: The model is not trained until perfect convergence. The idea of the unit-test is just to test the mechanics of the model, i.e. nothing crashes and all the functionality is available.

![learning_curves](https://github.com/JeffersonLab/jlab_datascience_core/assets/19531538/1922b6e7-9f1b-403a-bb6d-cab33c2068e6)

![reconstructed_data](https://github.com/JeffersonLab/jlab_datascience_core/assets/19531538/9947416b-7807-46bf-8909-4c77c8d430bb)

**Comment by dlersch** [comment]
Implemented a [residual analyzer analysis module](https://github.com/JeffersonLab/jlab_datascience_core/blob/5-evms-integrate-core-framework/jlab_datascience_toolkit/analysis/residual_analyzer.py). All this module does is to compute the difference between the prediction of a model and its input data:
```
residual = reduction_op[x - model(x),axis=reduction_axis]
```
where `reduction_op` is one of the following operations: mean (default), absolute value, squared value. The `reduction_axis` defines along which axis the reduction operation shall be performed. Since this project is predominantly dealing with images, the default is: `reduction_axis = 3`. Both, `reduction_op` and `reduction_axis` can be adjusted via a config file. The residual analyzer comes with its own [unit-test](https://github.com/JeffersonLab/jlab_datascience_core/blob/5-evms-integrate-core-framework/utests/utest_residual_analyzer.py) where the MNIST data is folded with a gaussian distribution. The original and folded data (aka reconstructed data) are then fed into the residual analyzer which produces diagnostic plots, such as the two shown below:

![res_0](https://github.com/JeffersonLab/jlab_datascience_core/assets/19531538/c1dd414d-7b1a-460b-9a29-7a6c15b5879d)

![res_11](https://github.com/JeffersonLab/jlab_datascience_core/assets/19531538/742f90a3-dd2b-405e-8848-616bc4c5be41)

**Comment by dlersch** [comment]
### Entire Worflow / Driver
Started to implement the driver. Using @sgoldenCS graph module.

**Comment by dlersch** [comment]
### Analysis: Data Reconstruction
Added a [data reconstruction module](https://github.com/JeffersonLab/jlab_datascience_core/blob/5-evms-integrate-core-framework/jlab_datascience_toolkit/analysis/data_reconstruction.py) which just takes a model m, input data x and collects the output:
```
y = m(x)
```
The data is handled via [tensorflows dataset](https://www.tensorflow.org/api_docs/python/tf/data/Dataset). This allows to handle large data sets, but also adds another level of flexibility to the workflow. The data reconstruction module allows to track the original data as well (in case the user decides to process only a small fraction of x). The data reconstruction analysis module was tested via its [unit-test](https://github.com/JeffersonLab/jlab_datascience_core/blob/5-evms-integrate-core-framework/utests/utest_data_reconstruction.py). 

The idea behind this module is that it can be combined with any other analysis module that requires a set of reconstructed data samples.

**Comment by dlersch** [comment]
### Analysis: Learning Curve Visualizer
Added analysis module which [visualizes learning curves](https://github.com/JeffersonLab/jlab_datascience_core/blob/5-evms-integrate-core-framework/jlab_datascience_toolkit/analysis/learning_curve_visualizer.py). The input for this module is a dictionary, such as  [tensorflows history object](https://www.tensorflow.org/api_docs/python/tf/keras/callbacks/History). This modul enables the user to produces as many plots as needed and to visualize all metrics that are recorded duirng the training of a model. This analysis module comes with a [dedicated unit-test](https://github.com/JeffersonLab/jlab_datascience_core/blob/5-evms-integrate-core-framework/utests/utest_learning_curve_visualizer.py), in which a dictionary with mock-data is passed to the analyzer. The mock data consists of four subsets that are separated into two groups which are then individually visualized. 

Shown below are the outputs of the unit-test.

![first_plot](https://github.com/JeffersonLab/jlab_datascience_core/assets/19531538/a591fb47-faec-4867-8df1-0f3fd2299d92)

![second_plot](https://github.com/JeffersonLab/jlab_datascience_core/assets/19531538/98734aca-6e15-4c86-bd6a-7ac1fb2a6892)

**Comment by dlersch** [comment]
### Data Parser: MNIST Data Parser
Add [MNIST data parsing module](https://github.com/JeffersonLab/jlab_datascience_core/blob/5-evms-integrate-core-framework/jlab_datascience_toolkit/data_parser/mnist_data_parser.py) which does not require any inputs, but produces and output which is simply the MNIST digits. The idea behind this parser is, that it can be used to test and debug a workflow or just a specific module. Since this parser does not require any user inputs it can be used right away for testing or a small analysis.  Furthermore will this module be helpful for quick demonstrations. This module has a dedicated [unit-test](https://github.com/JeffersonLab/jlab_datascience_core/blob/5-evms-integrate-core-framework/utests/utest_mnist_data_parser.py) which loads the MNIST data via said parser and visualizes it. Shown below is the example output of such a unit-test.

![mnist_data_plots](https://github.com/JeffersonLab/jlab_datascience_core/assets/19531538/bff9a384-3045-4179-9cd2-eac08cf7f2b0)

**Comment by dlersch** [comment]
### Image Anomaly Detection Workflow
As mentioned last week, I am using the graph building tool written by @sgoldenCS  to define a [image anomaly detection driver](https://github.com/JeffersonLab/jlab_datascience_core/blob/5-evms-integrate-core-framework/jlab_datascience_toolkit/driver/image_anomaly_detection.py). This driver uses all components that have been introduced in the previous posts. The driver consists of the following steps:
1. Use [MNIST data parser](https://github.com/JeffersonLab/jlab_datascience_core/blob/5-evms-integrate-core-framework/jlab_datascience_toolkit/data_parser/mnist_data_parser.py) to load MNIST digits
2. Scale MNIST data via the [numpy linear scaler](https://github.com/JeffersonLab/jlab_datascience_core/blob/5-evms-integrate-core-framework/jlab_datascience_toolkit/data_prep/numpy_linear_scaler.py)
3. Train a [keras CNN AE](https://github.com/JeffersonLab/jlab_datascience_core/blob/5-evms-integrate-core-framework/jlab_datascience_toolkit/model/keras_cnn_ae.py) on the scaled MNIST data
4. Use the [learning curve visualizer](https://github.com/JeffersonLab/jlab_datascience_core/blob/5-evms-integrate-core-framework/jlab_datascience_toolkit/analysis/learning_curve_visualizer.py) to plot the training loss curves
5. Feed the scaled MNIST data and the trained AE into the [data reconstruction module](https://github.com/JeffersonLab/jlab_datascience_core/blob/5-evms-integrate-core-framework/jlab_datascience_toolkit/analysis/data_reconstruction.py)
6. Re-scale the reconstructed data back via the  [numpy linear scaler](https://github.com/JeffersonLab/jlab_datascience_core/blob/5-evms-integrate-core-framework/jlab_datascience_toolkit/data_prep/numpy_linear_scaler.py)
7. Finally use the [residual analyzer](https://github.com/JeffersonLab/jlab_datascience_core/blob/5-evms-integrate-core-framework/jlab_datascience_toolkit/analysis/residual_analyzer.py) on the re-scaled data and plot residuals, i.e. the difference between the true and reconstructed MNIST digits.

Lines 14 - 23 nicely demonstrate how the above 7 steps define a sequential graph  in the [anomaly detection driver](https://github.com/JeffersonLab/jlab_datascience_core/blob/5-evms-integrate-core-framework/jlab_datascience_toolkit/driver/image_anomaly_detection.py) 

Shown below are the results produced by the anomaly detection driver. Please note, that the AE was not tuned at all. I used best guess values to train the model.

![losses](https://github.com/JeffersonLab/jlab_datascience_core/assets/19531538/69c92721-c542-489c-82ea-22e2f6911ef7)

![res_5](https://github.com/JeffersonLab/jlab_datascience_core/assets/19531538/851d34cb-c92b-4fc0-84c2-e9fb6bc6aade)

![res_2](https://github.com/JeffersonLab/jlab_datascience_core/assets/19531538/e0270196-1f50-448d-af55-1a2aeba73de0)

**Comment by dlersch** [comment]
### Hyper Parameter Optimization for Keras CNN AE
Implemented a [hyper parameter tuner for the keras CNN AE](https://github.com/JeffersonLab/jlab_datascience_core/tree/5-evms-integrate-core-framework/jlab_datascience_toolkit/hyper_parameter_tuning). The tuner is based on the [optuna framework](https://optuna.org/). The key idea behind this implementation is, that the hyper parameter tuner inherits from the [model core](https://github.com/JeffersonLab/jlab_datascience_core/blob/5-evms-integrate-core-framework/jlab_datascience_toolkit/core/jdst_model.py) and is therefore its own model module. Consequently, this model is also registered as a model. The registration for the keras CNN AE hyper parameter tuner can be found [here](https://github.com/JeffersonLab/jlab_datascience_core/blob/5-evms-integrate-core-framework/jlab_datascience_toolkit/model/__init__.py). This implementation comes with its own [default configuration](https://github.com/JeffersonLab/jlab_datascience_core/blob/5-evms-integrate-core-framework/jlab_datascience_toolkit/cfgs/defaults/hpo_keras_cnn_ae_cfg.yaml) and [unit-test](https://github.com/JeffersonLab/jlab_datascience_core/blob/5-evms-integrate-core-framework/utests/utest_hpo_keras_cnn_ae.py). The unit-test runs a very short and basic analysis. Shown below are a few results reported by the unit-test. The keras CNN AE was tuned for 10 trials on the MNIST data and the final / best model was trained for 20 epochs.
 
![optimization_history](https://github.com/user-attachments/assets/7c3c0335-9d1d-412e-8f15-e8e11bfa9ef0)

![learning_curves](https://github.com/user-attachments/assets/d8bcb6a8-528d-4c4b-b2eb-2015140d44db)

![hp_importance](https://github.com/user-attachments/assets/b98bcc4c-ed49-4184-a3c0-f21694f415f5)

![hp_parallel](https://github.com/user-attachments/assets/46c5bf92-889e-42d7-9742-0981d3cbb739)
